contrib/profiler/profiler.py

```diff
@@ -188,148 +143,188 @@ def get_network_bytes(filename):
-    for i in range(1, count):
-        if sample_list[i].get_cpu_usage() < min_cpu:
-            min_cpu = sample_list[i].get_cpu_usage()
-            min_cpu_idx = i
-        if sample_list[i].get_cpu_usage() > max_cpu:
-            max_cpu = sample_list[i].get_cpu_usage()
-            max_cpu_idx = i
-        if sample_list[i].get_gpu_usage() < min_gpu:
-            min_gpu = sample_list[i].get_gpu_usage()
-            min_gpu_idx = i
-        if sample_list[i].get_gpu_usage() > max_gpu:
-            max_gpu = sample_list[i].get_gpu_usage()
-            max_gpu_idx = i
-        if sample_list[i].get_mem_used() < min_mem:
-            min_mem = sample_list[i].get_mem_used()
-            min_mem_idx = i
-        if sample_list[i].get_mem_used() > max_mem:
-            max_mem = sample_list[i].get_mem_used()
-            max_mem_idx = i
-        if sample_list[i].get_read_bytes() > max_read:
-            max_read = sample_list[i].get_read_bytes()
-            max_read_idx = i
-
-        if sample_list[i].get_gpu_usage() <= 10:
-            cpu_when_gpu_low.append(sample_list[i].get_cpu_usage())
-            mem_when_gpu_low.append(sample_list[i].get_mem_used())
-            disk_read_when_gpu_low.append(sample_list[i].get_read_bytes())
-
-        sum_cpu += sample_list[i].get_cpu_usage()
-        sum_gpu += sample_list[i].get_gpu_usage()
-        sum_mem += sample_list[i].get_mem_used()
-        sum_read += sample_list[i].get_read_bytes()
-
-    length_gpu_low = len(cpu_when_gpu_low)
-    if length_gpu_low == 0:
-        length_gpu_low = 1
-
-    print('%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f\t%f' %
-          (max_gpu, sum_gpu / count, max_cpu, sum_cpu / count, max_mem, sum_mem / count, max_read, sum_read,
-           sum(cpu_when_gpu_low) / length_gpu_low,
-           sum(mem_when_gpu_low) / length_gpu_low,
-           sum(disk_read_when_gpu_low) / length_gpu_low))
-    return [max_gpu, sum_gpu / count, max_cpu, sum_cpu / count, max_mem, sum_mem / count, max_read, sum_read,
-            sum(cpu_when_gpu_low) / length_gpu_low,
-            sum(mem_when_gpu_low) / length_gpu_low,
-            sum(disk_read_when_gpu_low) / length_gpu_low]
+def analyze_samples(sample_list, advise):
+    sample_list = np.array(sample_list, dtype=np.float)
+    used_gpu_num = (sample_list.shape[1] - 7) / 4
+    cpu_usage = sample_list[:, 0]
+    mem_usage = sample_list[:, 1] / sample_list[:, 2]
+    gpu_usage = list()
+    gpu_mem_usage = list()
+    for i in range(int(used_gpu_num)):
+        gpu_usage.append(sample_list[:, 7 + i * 4])
+        gpu_mem_usage.append(sample_list[:, 9 + i * 4] / sample_list[:, 10 + i * 4])
+
+    if used_gpu_num >= 2:
+        # multiple GPUs, analyze whether each GPU has the same memory usage
+        gpu_mem_avg_0 = np.average(gpu_mem_usage[0])
+        gpu_mem_avg_1 = np.average(gpu_mem_usage[1])
+        if abs(gpu_mem_avg_0 - gpu_mem_avg_1) > 0.15:
+            advise.add_times(index=1)
+
+    if np.average(gpu_usage[0]) >= 90:
```

Comment:
should be 0.90?

Reply:
90% is from the testing of our model. The value can be changed if we can get more testing data.

Reply:
I mean for `gpu_mem_usage`, you use threshold as 0.80. But here you use threshold as 90. The unit is different?

---
In reply to: [332371115](https://github.com/microsoft/pai/pull/3649#discussion_r332371115) [](ancestors = 332371115)

Reply:
Yes, the 'gpu_mem_usage' is calculated by myself, and the value if from 0 to 1.
But the 'gpu_usage'  is gotten from the NVIDIA's API, it returns the numerator of the percent.

Reply:
Got it, thanks

---
In reply to: [332808207](https://github.com/microsoft/pai/pull/3649#discussion_r332808207) [](ancestors = 332808207)